doc/source/ray-core/examples/batch_prediction.ipynb

```diff
@@ -0,0 +1,283 @@
+    "* ML model: this is a trained ML model which is usually also stored in an external storage system.\n",
+    "* Predictions: these are the outputs when applying the ML model on observations. The predictions are usually written back to the storage system.\n",
+    "\n",
+    "With Ray, you can build scalable batch prediction for large datasets at high prediction throughput. This is achieved by splitting the dataset into disjoint shards and executing them in parallel, with either Ray tasks or Ray actors across a Ray cluster.\n",
+    "\n",
+    "```{tip}\n",
+    "In general, actor-based batch prediction is preferred for higher efficiency. This is because the actor is stateful and it can reuse the expensive initialization across tasks.\n",
+    "```\n",
+    "\n",
+    "## Task-based batch prediction\n",
+    "\n",
+    "With Ray tasks, you can build a batch prediction program in this way:\n",
+    "1. Loads the model\n",
+    "2. Launches Ray tasks, with each taking in the model and a shard of input dataset\n",
+    "3. Each worker executes predictions on the assigned shard, and writes out results\n",
+    "\n",
+    "Let’s take NYC taxi data in 2009 for example. Suppose we have this simple model:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pandas as pd\n",
+    "import numpy as np\n",
+    "\n",
+    "def load_model():\n",
+    "    # A dummy model.\n",
+    "    def model(batch: pd.DataFrame) -> pd.DataFrame:\n",
+    "        model.payload = np.zeros(100_000_000)\n",
+    "        return pd.DataFrame({\"score\": batch[\"passenger_count\"] % 2 == 0})\n",
+    "    \n",
+    "    return model"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The dataset has 12 files (one for each month) so we can naturally have each Ray task to take one file. By taking the model and a shard of input dataset (i.e. a single file), we can define a Ray remote task for prediction:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pyarrow.parquet as pq\n",
+    "import ray\n",
+    "\n",
+    "@ray.remote\n",
+    "def make_prediction(model, shard_path):\n",
+    "    df = pq.read_table(shard_path).to_pandas()\n",
+    "    result = model(df)\n",
+    "    # Write the result out.\n",
+    "    # Here we just return the size about the result.\n",
+    "    return len(result)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The driver launches all tasks for the entire input dataset. In order to not overload cluster and cause OOM, we use ray.wait() to limit the number of in-flight tasks, see details about this design pattern in [Using ray.wait() to limit tasks]] (https://docs.ray.io/en/master/ray-core/tasks/patterns/limit-tasks.html)"
```

Comment:
Is this what we are doing in dataset: using ray.wait to limit the number of in-flight tasks?

Reply:
Datasets is actually using auto-scaling actor pool.